server/schedule/operator_controller.go

```diff
@@ -103,53 +103,74 @@ func (oc *OperatorController) Dispatch(region *core.RegionInfo, source string) {
 		failpoint.Inject("concurrentRemoveOperator", func() {
 			time.Sleep(500 * time.Millisecond)
 		})
-		timeout := op.IsTimeout()
-		if step := op.Check(region); step != nil && !timeout {
-			operatorCounter.WithLabelValues(op.Desc(), "check").Inc()
 
-			// When the "source" is heartbeat, the region may have a newer
-			// confver than the region that the operator holds. In this case,
-			// the operator is stale, and will not be executed even we would
-			// have sent it to TiKV servers. Here, we just cancel it.
-			origin := op.RegionEpoch()
-			latest := region.GetRegionEpoch()
-			changes := latest.GetConfVer() - origin.GetConfVer()
-			if source == DispatchFromHeartBeat &&
-				changes > uint64(op.ConfVerChanged(region)) {
-
-				if oc.RemoveOperator(op) {
-					log.Info("stale operator",
-						zap.Uint64("region-id", region.GetID()),
-						zap.Duration("takes", op.RunningTime()),
-						zap.Reflect("operator", op),
-						zap.Reflect("latest-epoch", region.GetRegionEpoch()),
-						zap.Uint64("diff", changes),
-					)
-					operatorCounter.WithLabelValues(op.Desc(), "stale").Inc()
-					oc.opRecords.Put(op, pdpb.OperatorStatus_CANCEL)
-					oc.PromoteWaitingOperator()
-				}
-
-				return
-			}
+		// Update operator status:
+		// The operator status should be STARTED.
+		step := op.Check(region)
+		_ = op.CheckSuccess()
+		_ = op.CheckTimeout()
 
-			oc.SendScheduleCommand(region, step, source)
+		switch op.Status() {
+		case operator.STARTED:
+			operatorCounter.WithLabelValues(op.Desc(), "check").Inc()
+			if source != DispatchFromHeartBeat || !oc.checkStaleOperator(op, region) {
+				oc.SendScheduleCommand(region, step, source)
+			}
 			return
+		case operator.SUCCESS:
+			if oc.RemoveOperator(op) {
+				oc.PromoteWaitingOperator()
+			}
+		case operator.TIMEOUT:
+			if oc.RemoveOperator(op) {
+				oc.PromoteWaitingOperator()
+			}
+		default:
+			if oc.removeOperatorWithoutBury(op) {
+				// CREATED, EXPIRED must not appear.
+				// CANCELED, REPLACED must remove before transition.
+				log.Error("dispatching operator with unexpected status",
+					zap.Uint64("region-id", op.RegionID()),
+					zap.String("status", operator.OpStatusToString(op.Status())),
+					zap.Reflect("operator", op))
+				operatorCounter.WithLabelValues(op.Desc(), "unexpected").Inc()
+				failpoint.Inject("unexpectedOperator", func() {
+					panic(op)
+				})
+				_ = op.Cancel()
```

Comment:
IMO, we should check the return value as much as possible.

Reply:
`Cancel()` here is only used to ensure the operator status will become an end status, so the return value can be ignored. I use the underscore to explicitly ignore the return values that do not need to handle.